### PR TITLE
refactor: data scratch ios

### LINKED ISSUE #1755: Data scratch - ios test matrix
# Plan

According to the specification described in epic [#1727](https://github.com/Flank/flank/issues/1727)
add abstraction and implementation for the following part of the data layer:

## Abstraction

`ftl/data/TestMatrixIos.kt`

```kotlin
package ftl.data

val executeTestMatrixIos: TestMatrixIos.Execute get() = TODO()

object TestMatrixIos {

    data class Config(
        // args
        val clientDetails: Map<String, String>?,
        val networkProfile: String?,
        val directoriesToPull: List<String>,
        val testTimeout: String,
        val recordVideo: Boolean,
        val flakyTestAttempts: Int,
        val failFast: Boolean,
        val project: String,
        val resultsHistoryName: String?,

        // build
        val devices: List<Device>,
        val otherFiles: Map<String, String>,
        val additionalIpasGcsPaths: List<String>,
    )

    sealed class Type {
        data class XcTest(
            val xcTestGcsPath: String,
            val xcTestRunFileGcsPath: String,
            val xcodeVersion: String,
            val testSpecialEntitlements: Boolean,
            val matrixGcsPath: String,
        )

        data class GameLoop(
            val appGcsPath: String,
            val scenarios: List<Int>,
            val matrixGcsPath: String,
        )
    }

    interface Execute : (Config, Type) -> TestMatrix.Result
}
```

## Target

* `IosRunCommand` -> `IosArgs/runIosTests` -> `GcIosTestMatrix/build`
* `IosTestContext`

## Adapter

`ftl/adapter/TestMatrixIos.kt`

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #1755 

- Code is refactored according to description provided by #1755
- It takes into account work done on: https://github.com/Flank/flank/pull/1941 and attempts to be as close as possible for consistency.

## Checklist

- [x] Unit tested
- [x] Integration tests updated

### CODE REVIEW

**Comment by adamfilipow92** [comment]
@flank-it

File: test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+): IosTestSetup = IosTestSetup()
+    .setNetworkProfile(config.networkProfile)
+    .setPushFiles(config.otherFiles.mapToIosDeviceFiles())
+    .setAdditionalIpas(config.additionalIpasGcsPaths.mapGcsPathsToFileReference())
+    .setPullDirectories(config.directoriesToPull.toIosDeviceFiles())
+
+private fun List<String>.toIosDeviceFiles() = map { path -> toIosDeviceFile(path) }
+
+private val TestMatrixIos.Config.clientInfo
+    get() = ClientInfo().setName("Flank").setClientInfoDetails(clientDetails?.toClientInfoDetailList())
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt:93]: some of these functions are common for Android and iOS could we share them?

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt:93]: @piotradamczyk5 Sure can look into it :)

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/client/google/run/ios/GcIosTestMatrix.kt:93]: There is now differences in androids clientinfo. I dont see a reason to try merge it.

**Comment by Sloox** [comment]
@flank-it

**Comment by Sloox** [comment]
@flank-it

**Comment by pawelpasterz** [comment]
> https://gradle.com/s/2jpxa4zibqm5m

Tests failed due to quota limit, let's try again

**Comment by pawelpasterz** [comment]
@flank-it

**Comment by Sloox** [comment]
@flank-it

**Review by pawelpasterz — APPROVED** [review]

**Review by piotradamczyk5 — APPROVED** [review]

**Pull request merged**